src/python/pants/pantsd/service/scheduler_service.py

```diff
@@ -157,3 +157,4 @@ def run(self):
                 self._logger.critical(f"The scheduler was invalidated: {e!r}")
                 self.terminate()
         self._scheduler_session.cancel()
+        self._scheduler.shutdown()
```

Comment:
This is already executed here: https://github.com/pantsbuild/pants/blob/95850c9907fb84ee2b595a8bc8e043321c7d6953/src/python/pants/pantsd/pants_daemon.py#L199-L202

Reply:
What about when `--no-pantsd` is in effect?

Reply:
Good point. But `SchedulerService` isn't used in that case: that would be `LocalPantsRunner` without a session passed in: https://github.com/pantsbuild/pants/blob/cace8518a43646ac0e84c4f841cee01e46b1defc/src/python/pants/bin/local_pants_runner.py#L60-L81

That same codepath is used in both the daemon and no-daemon cases, so you'd need to essentially differentiate between an "owned" `GraphSession` (which was created for a run and needs to be shut down, and a borrowed (from the daemon) `GraphSession`.

Reply:
Understood. "That same codepath is used in both the daemon and no-daemon cases," is the bit of context that I was missing.